.circleci/config.yml

```diff
@@ -1,109 +1,359 @@
-version: 2
-jobs:
-  build:
-    parallelism: 6
-    working_directory: ~/wp-calypso
-    docker:
-      - image: circleci/node:10.10.0-browsers
-        environment:
-          CIRCLE_ARTIFACTS: /tmp/artifacts
-          CIRCLE_TEST_REPORTS: /tmp/test_results
+references:
+  shared-environment: &shared-environment
+    CIRCLE_ARTIFACTS: /tmp/artifacts
+    CIRCLE_TEST_REPORTS: /tmp/test_results
 
-    steps:
-      - restore_cache:
-          name: Restore git cache
-          key: v1-20180524-git
-
-      - checkout
-
-    # Update origin/master so we can diff against it later
-      - run:
-          name: Update master branch
-          command: git fetch --force origin master
+  docker-image: &docker-image
+    - image: circleci/node:10.10.0-browsers
 
-      - save_cache:
-          name: Save git cache
-          key: v1-20180524-git
-          paths:
-            - ".git"
+  setup-results-and-artifacts: &setup-results-and-artifacts
+    name: Create Directories for Results and Artifacts
+    command: |
+      mkdir -p                             \
+        "$CIRCLE_ARTIFACTS/translate"      \
+        "$CIRCLE_TEST_REPORTS/client"      \
+        "$CIRCLE_TEST_REPORTS/eslint"      \
+        "$CIRCLE_TEST_REPORTS/integration" \
+        "$CIRCLE_TEST_REPORTS/server"      \
+        "$HOME/jest-cache"
 
-      - run:
-          name: Create Directories for Results and Artifacts
-          command: |
-            mkdir -p $CIRCLE_ARTIFACTS
-            mkdir -p $CIRCLE_TEST_REPORTS
+  # Jest cache caching
+  #
+  # Jest uses a cache to speed up builds. If we persist this cache across builds,
+  # we can improve the speed of subsequent builds.
+  #
+  # Circle caches never overwritten, so we must ensure that Jest caches from different jobs
+  # do not collide or we'll only cache 1 job.
+  #
+  # We also need to ensure that different nodes and different total nodes do not collide.
+  # When we split tests, different nodes will receive a different set of tests so each node's
+  # cache should be unique.
+  #
+  # Finally, we cache on the branch and revision, falling back to master. This should give us
+  # pretty good "nearest neighbor" primer for the Jest cache.
+  #
+  # More about the CircleCI cache: https://circleci.com/docs/2.0/caching
+  restore-jest-cache: &restore-jest-cache
+    name: Restore Jest cache
+    keys:
+      - v7-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}-{{ .Branch }}-{{ .Revision }}
+      - v7-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}-{{ .Branch }}
```

Comment:
since branches can be rebased and can change will this cause cache pollution? in other words, should be caching on the commit hash instead of the branch head?

Reply:
I wouldn't expect it to cause problems. This is warming the Jest cache. It will try to match a cache from the exact revision, then the most recent from the branch, then the most recent from master…

The most recent cache from the same branch is probably a good guess. In the case of the first commit on an old branch after rebasing on master, it would be better to match master, but I'd expect that to be a less common case.

I'm not completely certain, but I don't think there's any danger here. We just have more or less warm caches. I expect the same situation arises with local test runs. After rebasing, Jest will have a cache based on the older branch point. That doesn't mean the tests become inaccurate, just that Jest needs to do more work than with a very recent cache.

Related: https://github.com/Automattic/wp-calypso/pull/27180#discussion_r218834521

Reply:
I'm thinking about this for all these `.branch` caches as well. Just to confirm:

 - I create a branch and push code to it
 - CircleCI writes to these caches for my branch
 - I push a new commit to my branch (not even rebasing)
 - CircleCI reuses the old cache

Or does CircleCI even use the branch tip when it uses `.Branch` - I guess it's possible that the syntax is shorthand for "current branch tip" vs. "current branch name"

Reply:
> * I create a branch and push code to it
> * CircleCI writes to these caches for my branch
> * I push a new commit to my branch (not even rebasing)
> * CircleCI reuses the old cache

Yes and no 🙃 These cache restores will cascade, matching the first that hits, top to bottom. In this case, on the first build:

- Attempt restore `…-{{ .Branch }}-{{ .Revision }}`. Cache miss
- Attempt restore `…-{{ .Branch }}`. All revisions from this branch hit. The most recent cache from this branch will be restored.
- Jest runs tests…
- `…-{{ .Branch }}-{{ .Revision }}` New Jest cache cache is stored.

On subsequent runs of this revision, we'll match the new Jest cache cache that matches exactly. Also, the next commit to this branch will match this newly created cache.

---

I don't understand the fundamental concern. We are not caching test results. Jest still runs, we only restore Jest's cache so that it's warm rather than cold. I would not expect test results to be inaccurate due to an older cache being provided for Jest. As I understand it, that's the purpose and expected behavior of a cache like the one Jest uses– to store some previous work that's already been done to avoid repeating it.

Reply:
According to the docs `.Revision` is `The VCS revision currently being built.` which I guess means commit hash…?

My fundamental concern is restoring files from a cache that have updated and then we "avoid repeating" those tests because we think they didn't change.

I'm only trying to make sure we don't skip tests due to restoring a cache from old code.

Reply:
> According to [the docs](https://circleci.com/docs/2.0/caching/) `.Revision` is `The VCS revision currently being built.` which I guess means commit hash…?

Yep 👍 

> My fundamental concern is restoring files from a cache that have updated and then we "avoid repeating" those tests because we think they didn't change.
> 
> I'm only trying to make sure we don't skip tests due to restoring a cache from old code.

I'm confident that is not happening. Jest does have options to skip tests based on changes, but we're not using any of them. I've pushed a commit to demonstrate how Jest fails correctly with this caching scheme: https://circleci.com/workflow-run/8de0cfa1-4ae0-45b6-bd77-30c607012907

Reply:
your confidence gives me confidence 😄